Exec task fails in Windows Nano Server docker containers using MSBuild 15.3

Something in the Exec task is failing to launch basic command line processes in MSBuild 15.3, but only when running in Nano Server docker containers. This worked fine in MSBuild 15.1.

**Repro**
Install [Docker for Windows](https://store.docker.com/editions/community/docker-ce-desktop-windows) and ensure "Windows containers" are used (not Linux containers).

Create a dockerfile like this and a file named `test.proj`.
```Dockerfile
FROM microsoft/dotnet-nightly:2.0.0-preview3-sdk-nanoserver
WORKDIR /app
COPY test.proj .
RUN dotnet msbuild test.proj
```
```xml
<!-- test.proj -->
<Project>
    <Target Name="Build">
        <Exec Command="dir" />
    </Target>
</Project>
```

Execute
```
docker build .
```

**Expected**
Build target runs and executed the "dir" command.

**Actual**
The `dotnet msbuild test.proj` command fails with:
```
C:\app\test.proj(3,9): error MSB6004: The specified task executable location "C:\Users\ContainerAdministrator\cmd.exe" is invalid.
```

**Details**

MSBuild version: 15.3.246.41955 for .NET Core
Docker base image: `microsoft/nanoserver:10.0.14393.1358`
OS: Windows 10
Docker: 17.06
Environment variables inside the container:
```
PATH = C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Users\ContainerAdministrator\AppData\Local\Microsoft\WindowsApps;C:\Program Files\dotnet;
USERPROFILE = C:\Users\ContainerAdministrator
```

You can gather logs by launching a container with a volume mount to the folder containing the test.proj file.
```
docker run --rm -v "$(get-location):C:/app" -it -w C:/app microsoft/dotnet-nightly:2-sdk
C:\app> dotnet msbuild test.proj /bl
```

cmd.exe itself runs run, and is located in `C:\Windows\system32\cmd.exe`.
There is no file named `C:\Users\ContainerAdministrator\cmd.exe`

Full diagnostic log:
[msbuild-2-sdk.txt](https://github.com/Microsoft/msbuild/files/1136947/msbuild-2-sdk.txt)

**For comparision**
This works just fine using MSBuild 15.1. Example
```Dockerfile
FROM microsoft/dotnet:1-sdk

WORKDIR /app
COPY test.proj .
RUN dotnet msbuild test.proj
```
[msbuild-1-sdk.txt](https://github.com/Microsoft/msbuild/files/1136953/msbuild-1-sdk.txt)

I'm less sure now. We have our own implementation of the locate-the-special-folder code.

I made this local change:

```diff
diff --git a/src/Shared/FileUtilities.GetFolderPath.cs b/src/Shared/FileUtilities.GetFolderPath.cs
index 315354fc..613cafbd 100644
--- a/src/Shared/FileUtilities.GetFolderPath.cs
+++ b/src/Shared/FileUtilities.GetFolderPath.cs
@@ -35,6 +35,8 @@ namespace Microsoft.Build.Shared
                                                       Win32Native.SHGFP_TYPE_CURRENT, /* dwFlags:   [in] retrieve current path */
                                                       sb);                            /* pszPath:   [out]resultant path */
 
+            Console.WriteLine($"ShGetFolderPath for {folder} returned {hresult} and gave up {sb.ToString()}");
+
             String s;
             if (hresult < 0)
             {
diff --git a/src/Utilities/ToolTask.cs b/src/Utilities/ToolTask.cs
index 38dcad86..985ccc6b 100644
--- a/src/Utilities/ToolTask.cs
+++ b/src/Utilities/ToolTask.cs
@@ -619,6 +619,8 @@ namespace Microsoft.Build.Utilities
                 // Otherwise, try to find the tool ourselves.
                 pathToTool = GenerateFullPathToTool();
 
+                Console.WriteLine($"** {pathToTool}");
+
                 // We have no toolpath, but we have been given an override
                 // for the tool exe, fix up the path, assuming that the tool is in the same location
                 if (pathToTool != null && !String.IsNullOrEmpty(_toolExe))
```

And get this output when plunking current 15.3 bits on top of CLI 1.0.4: 

```
ShGetFolderPath for LocalApplicationData returned 0 and gave up C:\Users\ContainerAdministrator
ShGetFolderPath for LocalApplicationData returned -2147024693 and gave up C:\Users\ContainerAdministrator
ShGetFolderPath for ApplicationData returned 0 and gave up C:\Users\ContainerAdministrator
ShGetFolderPath for ProgramFiles returned -2147024693 and gave up C:\Users\ContainerAdministrator
ShGetFolderPath for ProgramFilesX86 returned 0 and gave up C:\Users\ContainerAdministrator
ShGetFolderPath for System returned 0 and gave up C:\Users\ContainerAdministrator
** C:\Users\ContainerAdministrator\cmd.exe
C:\app\test.proj(3,9): error MSB6004: The specified task executable location "C:\Users\ContainerAdministrator\cmd.exe" is invalid.
```

That should isolate against the runtime, but it's still failing.

--- a/src/Tasks/Exec.cs
+++ b/src/Tasks/Exec.cs
@@ -520,6 +520,11 @@ namespace Microsoft.Build.Tasks
         /// <returns>path to cmd.exe</returns>
         protected override string GenerateFullPathToTool()
         {
+            return CommandProcessorPath.Value;
+        }
+
+        private static readonly Lazy<string> CommandProcessorPath = new Lazy<string>(() =>
+        {
             // Get the fully qualified path to cmd.exe
             if (NativeMethodsShared.IsWindows)
             {
@@ -541,7 +546,7 @@ namespace Microsoft.Build.Tasks
             {
                 return "sh";
             }
-        }
+        });
 
         /// <summary>
         /// Gets the working directory to use for the process. Should return null if ToolTask should use the

--- a/src/Tasks/Exec.cs
+++ b/src/Tasks/Exec.cs
@@ -523,7 +523,19 @@ namespace Microsoft.Build.Tasks
             // Get the fully qualified path to cmd.exe
             if (NativeMethodsShared.IsWindows)
             {
-                return ToolLocationHelper.GetPathToSystemFile("cmd.exe");
+                var systemCmd = ToolLocationHelper.GetPathToSystemFile("cmd.exe");
+
+#if !FEATURE_SPECIAL_FOLDERS
+                // Work around https://github.com/Microsoft/msbuild/issues/2273 and
+                // https://github.com/dotnet/corefx/issues/19110, which result in
+                // a bad path being returned above on Nano Server SKUs of Windows.
+                if (!File.Exists(systemCmd))
+                {
+                    return Environment.GetEnvironmentVariable("ComSpec");
+                }
+#endif
+
+                return systemCmd;
             }
             else
             {

